proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1269 @@
+  for an empty list ``[]ₑ`` or the list type constructor ``[]ₜ``.
+
+  With ``DataKinds``, the ``'[a]`` syntax in a type-level context is
+  resolved as ``[a]ₑ``; in a term-level context, this syntax is not
+  available.
+
+* The meaning of ``*`` has been determined. It can stand for one of the following:
+
+  1. ``Type`` from the ``Data.Kind`` module (under ``-XStarIsType``)
+  2. An occurrence of a term-level ``(*)`` infix operator
+  3. An occurrence of a type-level ``(*)`` infix operator (under ``-XTypeOperators``)
+
+* The meaning of ``'`` has been determined. It can stand for one of the following:
+
+  1. Namespace selection syntax (under ``-XDataKinds``)
+  2. Name quotation syntax (under ``-XTemplateHaskell``)
+
+Part II: Proposed Change Specification
+--------------------------------------
+
+Syntax
+~~~~~~
+
+1. Extend the term syntax with several constructs that
+   previously could only occur at the type level:
+
+   * Function arrows: ``a -> b``
+   * Multiplicity-polymorphic function arrows: ``a %m -> b`` (under ``-XLinearTypes``)
+   * Constraint arrows: ``a => b``
+   * Universal quantification: ``forall a. b``
+   * Visible universal quantification: ``forall a -> b``.
+
+   We will call them **types-in-terms**.
+
+   Grammatically, their constituents are terms, not types::
+
+                   proposed grammar:                      as opposed to:
+         ┌────────────────────────────────────┬───────────────────────────────────────┐
+         │                                    │                                       │
+         │  exp ::=                           │    exp ::=                            │
+         │      | exp₀ '->' exp₁              │        | type₀ '->' type₁             │
+         │      | exp₀ '=>' exp₁              │        | type₀ '=>' type₁             │
+         │      | 'forall' tv_bndrs '.'  exp  │        | 'forall' tv_bndrs '.'  type  │
+         │      | 'forall' tv_bndrs '->' exp  │        | 'forall' tv_bndrs '->' type  │
+         │                                    │                                       │
+         └────────────────────────────────────┴───────────────────────────────────────┘
+
+   This is a necessity to avoid parsing conflicts, with the following
+   consequences:
+
+   1. The ``'`` symbol signifies Template Haskell name quotation rather than ``DataKinds`` promotion.
+   2. The ``*`` symbol is treated as an infix operator regardless of ``-XStarIsType``.
+   3. Built-in syntax for tuples and lists is interpreted as in terms.
+      That is, ``[a]`` is a singleton list rather than the type of a list,
+      and ``(a, b)`` is a pair rather than the type of a pair.
+
+2. Make ``forall`` a keyword at the term level. Not guarded by any extension
+   (same motivation as `#193 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0193-forall-keyword.rst>`_).
+   This implies ``forall`` is no longer a valid identifier.
```

Comment:
In #193 it is said that the breakage of code that uses `forall` as type variable is probably ok, but shied from breaking code that uses `forall` as terms, and for good reason: There are a number of uses of that: https://hackage-search.serokell.io/?q=%5Eforall+%3A%3A

I have doubts that “good error messages” is good enough motivation to “break existing code in ways that can’t be fixed without changing APIs”. Can we reconsider this point maybe and hide `forall`-as-a-keyword-in-terms behind the extension as well? Or behind a `-XForallInTypes` extension that off in `Haskell2010` and `GHC2021`, but that we can turn on with `GHC2021+1` to nudge people away from that use, without forcing them.

(NB: Users of such a library can upgrade just fine, as `Foo.forall` works even if `forall` is a keyword.)

Reply:
On its surface, avoiding breaking changes like this sounds like a good idea, but in the end I think it will be a disservice to our users.

If you get an error because `forall` is now a keyword, you just rename the identifier to `forall_`. But if it’s sometimes a keyword, sometimes a valid name, and it depends on the compiler version, on the extensions enabled, and on whether the name is used qualified, then this is a complicated bit of trivia just to support a rarely used identifier.

Reply:
> Just to be clear, you’re suggesting -XForallInTypes as a migration strategy, not as an extension that we will support indefinitely, right?

(I should have written `-XForallInTerms` probably.)

Yes, this is just to make the migration less painful. With your proposal, the authors of library who _export_ a name called `forall` cannot upgrade to (say) GHC-9.2 at all without breaking all their clients. If it’s behind some extension, they can upgrade the compiler, and separately (and more slowly) start deprecating their `forall` export, and introduce a new one. We can make `-XForallInTerms` the _default_ quickly though (It wouldn’t affect Cabal packages that define a `default-language`). And removal of  support for `-XNoForallInTerms` can happen in the far future – that flag, although a wart, will not be horrible to support in the parser.

I am happy to disallow `-XNoForallInTerms` if `-XRequiredTypeArguments` is on; that should address that corner case.

Reply:
>  and separately (and more slowly) start deprecating their forall export, and introduce a new one.

What do we gain by making it a slow process?

Reply:
> What do we gain by making it a slow process?

We gain goodwill from otherwise possibly frustrated library maintainers, who can’t support newer GHCs without changing their APIs. to me “breaking the ability to provide a certain, existing, API” is qualitatively more severe as “breaking existing code”.

> How do you imagine it would work?

In the grammar for expressions, `forall` is only a keyword if `-XForallInTerms` is set – but maybe I am missing something here?

Reply:
> We gain goodwill from otherwise possibly frustrated library maintainers, who can’t support newer GHCs without changing their APIs.

I will reconsider my position when I hear one actual maintainer of a library with `forall` used as an identifier say that they would prefer a migration strategy that involves a new language extension.

You are probably underestimating the effect on the parser that this extension would have (see the next point).

> In the grammar for expressions, `forall` is only a keyword if `-XForallInTerms` is set – but maybe I am missing something here?

Yes, it misses that there’s no way of saying “in the grammar of expressions” when you’re lexing a token and deciding whether `forall` is a keyword or an identifier. And checking for extensions is done in the lexer, the grammar is not parameterized by the enabled extensions.

Reply:
I’m confused; isn't that the status quo that `forall` is a keyword in types, but not in terms (at least with `-XExplicitForAll` or related extensions on)?

Reply:
Hmm, that’s a bummer. I still think that “you cannot upgrade GHC unless you change your API” shouldn’t be done easily; I’ve raised it on the committee mailing list. Implementation complexity is of course a factor here.